pkg/services/query/query.go

```diff
@@ -213,6 +213,15 @@ func (s *Service) handleQuerySingleDatasource(ctx context.Context, user *user.Si
 		if cookieStr := parsedReq.httpRequest.Header.Get("Cookie"); cookieStr != "" {
 			req.Headers["Cookie"] = cookieStr
 		}
+
+		// pass through http headers for plugins to use
+		for k, v := range parsedReq.httpRequest.Header {
+			if len(v) == 0 {
+				req.Headers[k] = ""
+			} else {
+				req.Headers[k] = v[0] // assume only the first would be used
+			}
+		}
```

Comment:
First of all these changes is blocked by https://github.com/grafana/grafana/pull/58132, which refactors everything related to forwarding of HTTP headers, such as Cookies and OAuth token, and is targeted for Grafana v9.4.0.

Second, we don't want to do it like this because `req.Headers` is not meant for solely HTTP headers - it's purpose is environment info/custom metadata, see
https://github.com/grafana/grafana-plugin-sdk-go/blob/701ffb57ba2dd097fc765991cba38100ea41a41f/proto/backend.proto#L122-L123
https://github.com/grafana/grafana-plugin-sdk-go/blob/701ffb57ba2dd097fc765991cba38100ea41a41f/proto/backend.proto#L175-L176

We need to figure out how to allow HTTP headers to be set in this environment info. Here's a WIP/proposal of what this might look like: https://github.com/grafana/grafana-plugin-sdk-go/pull/562. I suggest that we continue this discussion in https://github.com/grafana/grafana/discussions/58287